src/counter.rs

```diff
@@ -12,37 +12,50 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use atomic64::{Atomic, AtomicF64, AtomicI64, PrimitiveNumber};
 use desc::Desc;
 use errors::Result;
 use metrics::{Collector, Metric, Opts};
 use proto;
 use std::collections::HashMap;
+use std::marker::PhantomData;
 use std::sync::Arc;
 use value::{Value, ValueType};
 use vec::{MetricVec, MetricVecBuilder};
 
 /// `Counter` is a Metric that represents a single numerical value that only ever
 /// goes up.
-#[derive(Clone)]
-pub struct Counter {
-    v: Arc<Value>,
+pub struct GenericCounter<P: Atomic> {
```

Comment:
em, why not using common template name T here?

Reply:
Here P is normally for something like AtomicF64, and there is a P::T for something like f64. To avoid confusion (i.e. what the hell is T::T) I keep this naming convention in this PR.